### PR TITLE
Add create and list MetricProfile APIs

### PR DESCRIPTION
**Pull request by shreyabiradar07**
## Description

This PR has following changes:

- Add `KruizeMetricProfileEntry` database table and creates metric profile when `local=true`
- Add `MetricProfileService` with `/createMetricProfile` and `/listMetricProfiles` REST APIs
- Modifies existing `PerformanceProfile` object to include  `apiVersion`, `kind` and `metadata` optional fields

### Type of change

- [ ] Bug fix
- [X] New feature
- [ ] Docs update
- [ ] Breaking change (What changes might users need to make in their application due to this PR?)
- [X] Requires DB changes

## How has this been tested?

Please describe the tests that were run to verify your changes and steps to reproduce. Please specify any test configuration required. 

- [ ] New Test X
- [ ] Functional testsuite

**Test Configuration**
* Kubernetes clusters tested on: minikube

## Checklist :dart:

- [X] Followed coding guidelines
- [X] Comments added
- [ ] Dependent changes merged
- [x] Documentation updated
- [ ] Tests added or updated

## Additional information

docker image - `quay.io/shbirada/metric_profile:apis`

### CODE REVIEW

**Comment by msvinaykumar** [comment]
@shreyabiradar07 
Explain the necessity of metric profiles and provide a brief guide on their utilization within the monitoring mode API documentation. Include references to the newly created markdown files.

File: design/MetricProfile.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# Create Metric Profile
+
+Client will invoke the Kruize createMetricProfile API for each deployment. Documentation still in progress stay
```
Comment (msvinaykumar) [inline design/MetricProfile.md:3]: Not for each deployment can you please refer this link and rephrase above accordingly
https://docs.google.com/document/d/12CHxowvyyOTUeBXLg2K4NeE7Q4PFuVO-pjHr4cgXWcs/edit

Comment (msvinaykumar) [reply design/MetricProfile.md:3]: Combine MetricProfile and MetricProfileAPI into a single file and include a reference in MonitotingModeAPI.md.

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfile.java [hunk 2]
```diff
@@ -17,6 +17,7 @@
 
 import com.autotune.analyzer.kruizeObject.SloInfo;
 import com.autotune.analyzer.recommendations.term.Terms;
+import com.fasterxml.jackson.databind.JsonNode;
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfile.java:20]: Before this pull request, as we discussed earlier, we need a straightforward PR that simply changes the name from PerformanceProfile to MetricProfile without altering the functionality. This current PR should build upon that.

**Comment by msvinaykumar** [comment]
Before this PR, as we discussed earlier, we need a straightforward PR that simply changes the name from PerformanceProfile to MetricProfile without altering the functionality. This current PR should build upon that.

File: src/main/java/com/autotune/utils/MetricsConfig.java [hunk 3]
```diff
@@ -56,6 +58,9 @@ private MetricsConfig() {
         timerBAddPerfProfileDB = Timer.builder("kruizeDB").description(DB_METRIC_DESC).tag("method", "addPerformanceProfileToDB");
         timerBLoadPerfProfileName = Timer.builder("kruizeDB").description(DB_METRIC_DESC).tag("method", "loadPerformanceProfileByName");
         timerBLoadAllPerfProfiles = Timer.builder("kruizeDB").description(DB_METRIC_DESC).tag("method", "loadAllPerformanceProfiles");
+        timerBAddMetricProfileDB = Timer.builder("kruizeDB").description(DB_METRIC_DESC).tag("method", "addMetricProfileToDB");
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/MetricsConfig.java:61]: Please exclude these metrics. There's no need to track them at the moment. If necessary, we can create a separate PR later.

Comment (shreyabiradar07) [reply src/main/java/com/autotune/utils/MetricsConfig.java]: done

**Comment by msvinaykumar** [comment]
Review Checklist

#### Common Review
- [ ] Code formatting adheres to the project’s style guide.
- [ ] No unnecessary commented-out code.
- [ ] Appropriate use of comments and documentation.

#### Logical Flow Review
- [ ] Code logic is clear and easy to follow.
- [ ] All branches and conditions are covered and tested.
- [ ] No redundant or duplicate code.

#### Demo Update
- [ ] Demonstration examples are updated to reflect changes.
- [ ] README and documentation include updated demo instructions.
- [ ] Any necessary demo files are included and tested.

#### Naming Conventions
- [ ] Variable and function names follow the project's naming conventions.
- [ ] File names are descriptive and adhere to naming standards.
- [ ] No ambiguous or misleading names.

#### Workflow
- [ ] CI/CD pipelines are updated and passing.
- [ ] Tests are included and all pass.
- [ ] The workflow for deploying or integrating changes is clear and documented.

#### Does It Affect ROS 
- [ ] Changes are reviewed for compatibility with ROS.
- [ ] ROS-specific configurations are updated if necessary.
- [ ] Any ROS dependencies or integrations are tested and verified.

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java [hunk 4]
```diff
@@ -302,6 +322,258 @@ public ValidationOutputData validateMandatoryFieldsAndData(PerformanceProfile pe
 
         return validationOutputData;
     }
+
+    /**
+     * Validates the data present in the metric profile object before adding it to the map
+     * @param metricProfile
+     * @return
+     */
+    private ValidationOutputData validateMetricProfileData(PerformanceProfile metricProfile) {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:331]: remove this function

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java [hunk 5]
```diff
@@ -80,6 +89,17 @@ public ValidationOutputData validate(PerformanceProfile performanceProfile) {
         return validatePerformanceProfileData(performanceProfile);
     }
 
+    /**
+     * Validates function variables
+     *
+     * @param metricProfile Metric Profile Object to be validated
+     * @return Returns the ValidationOutputData containing the response based on the validation
+     */
+    public ValidationOutputData validateMetricProfile(PerformanceProfile metricProfile) {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:98]: remove this function
and use validateMetricProfileData directly

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java [hunk 6]
```diff
@@ -302,6 +322,258 @@ public ValidationOutputData validateMandatoryFieldsAndData(PerformanceProfile pe
...
+     * @return
+     */
+    private ValidationOutputData validateMetricProfileData(PerformanceProfile metricProfile) {
+        // validate the mandatory values first
+        ValidationOutputData validationOutputData = validateMandatoryMetricProfileFieldsAndData(metricProfile);
+
+        // If the mandatory values are present,proceed for further validation else return the validation object directly
+        if (validationOutputData.isSuccess()) {
+            try {
+                new ExperimentDBService().loadAllMetricProfiles(performanceProfilesMap);
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:338]: Create a separate PR and avoid DB call rather make use of map

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java (hunk 4 above)
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:331]: can you merge some of the very common functionalities likes
 // Check if k8s type is supported
 SLO validations etc...  with
 validatePerformanceProfileData

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java [hunk 7]
```diff
@@ -302,6 +322,258 @@ public ValidationOutputData validateMandatoryFieldsAndData(PerformanceProfile pe
...
+    /**
+     * Validates the data present in the metric profile object before adding it to the map
+     * @param metricProfile
+     * @return
+     */
+    private ValidationOutputData validateMetricProfileData(PerformanceProfile metricProfile) {
+        // validate the mandatory values first
+        ValidationOutputData validationOutputData = validateMandatoryMetricProfileFieldsAndData(metricProfile);
+
+        // If the mandatory values are present,proceed for further validation else return the validation object directly
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:335]: enclose from here to end with try catch
and catch block add this
    validationOutputData.setSuccess(false);
                    validationOutputData.setMessage(errorString.toString());

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java (hunk 5 above)
Comment (shreyabiradar07) [reply src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:98]: currently `validateMetricProfile()` calls private method `validateMetricProfileData()` encapsulating the detailed validation logic, will it be okay to make it publicly accessible?

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java [hunk 8]
```diff
@@ -302,6 +322,258 @@ public ValidationOutputData validateMandatoryFieldsAndData(PerformanceProfile pe
 
         return validationOutputData;
     }
+
+    /**
+     * Validates the data present in the metric profile object before adding it to the map
+     * @param metricProfile
+     * @return
```
[the remark this answers is not in the document]
Comment (shreyabiradar07) [reply src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java]: updated docs

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java (hunk 4 above)
Comment (shreyabiradar07) [reply src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:229]: validation of common fields like `k8s_type` & `slo` is moved to `validateCommonProfileFields()`

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java (hunk 7 above)
Comment (shreyabiradar07) [reply src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java]: done

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java (hunk 5 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfileValidation.java:98]: yes

File: design/MetricProfile.md (hunk 1 above)
Comment (shreyabiradar07) [reply design/MetricProfile.md]: As discussed included MetricProfile reference in `KruizeLocalAPI.md` specific to local monitoring

**Review by msvinaykumar — APPROVED** [review]
LGTM


Review Checklist

#### Common Review
- [x] Code formatting adheres to the project’s style guide.
- [x] No unnecessary commented-out code.
- [x] Appropriate use of comments and documentation.

#### Logical Flow Review
- [x] Code logic is clear and easy to follow.
- [x] All branches and conditions are covered and tested.
- [x] No redundant or duplicate code.

#### Demo Update
- [x] Demonstration examples are updated to reflect changes.
- [x] README and documentation include updated demo instructions.
- [ ] Any necessary demo files are included and tested.

#### Naming Conventions
- [x] Variable and function names follow the project's naming conventions.
- [x] File names are descriptive and adhere to naming standards.
- [x] No ambiguous or misleading names.

#### Workflow
- [ ] CI/CD pipelines are updated and passing.
- [ ] Tests are included and all pass.
- [ ] The workflow for deploying or integrating changes is clear and documented.

#### Does It Affect ROS 
- [x] Changes are reviewed for compatibility with ROS.
- [x] ROS-specific configurations are updated if necessary.
- [x] Any ROS dependencies or integrations are tested and verified.

**Review by msvinaykumar — APPROVED** [review]
LGTM

**Pull request merged**